staging/src/k8s.io/apiserver/pkg/storage/testing/store_tests.go

```diff
@@ -619,6 +621,248 @@ func RunTestPreconditionalDeleteWithOnlySuggestionPass(ctx context.Context, t *t
+		pod := pods[i]
+		podOut := &example.Pod{}
+		err := store.Create(ctx, computePodKey(pod), pod, podOut, 0)
+		if err != nil {
+			panic(fmt.Sprintf("Unexpected error %s", err))
+		}
+		listOut := &example.PodList{}
+		err = store.GetList(ctx, "/pods", storage.ListOptions{
+			Recursive:            true,
+			ResourceVersion:      podOut.ResourceVersion,
+			ResourceVersionMatch: match,
+			Predicate: storage.SelectionPredicate{
+				Label: labels.Everything(),
+				Field: fields.Everything(),
+				Limit: 1,
+			},
+		}, listOut)
+		if err != nil {
+			panic(fmt.Sprintf("Unexpected error %s", err))
+		}
+		objectCount.Add(uint64(len(listOut.Items)))
+	}
+	b.ReportMetric(float64(objectCount.Load())/float64(b.N), "objects/op")
+}
+
+func RunBenchmarkStoreList(ctx context.Context, b *testing.B, store storage.Interface) {
+	namespaceCount := 100
+	podPerNamespaceCount := 100
+	var paginateLimit int64 = 100
+	nodeCount := 100
+	namespacedNames, nodeNames := prepareBenchchmarkData(ctx, store, namespaceCount, podPerNamespaceCount, nodeCount)
+	b.ResetTimer()
+	maxRevision := 1 + namespaceCount*podPerNamespaceCount
+	cases := []struct {
+		name  string
+		match metav1.ResourceVersionMatch
+	}{
+		{
+			name:  "RV=Empty",
+			match: "",
+		},
+		{
+			name:  "RV=NotOlderThan",
+			match: metav1.ResourceVersionMatchNotOlderThan,
+		},
+		{
+			name:  "RV=MatchExact",
+			match: metav1.ResourceVersionMatchExact,
+		},
+	}
+
+	for _, c := range cases {
+		b.Run(c.name, func(b *testing.B) {
+			runBenchmarkStoreList(ctx, b, store, 0, maxRevision, c.match, false, nodeNames)
+		})
+	}
+	b.Run("Paginate", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreList(ctx, b, store, paginateLimit, maxRevision, c.match, false, nodeNames)
+			})
+		}
+	})
+	b.Run("NodeIndexed", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreList(ctx, b, store, 0, maxRevision, c.match, true, nodeNames)
+			})
+		}
+		b.Run("Paginate", func(b *testing.B) {
+			for _, c := range cases {
+				b.Run(c.name, func(b *testing.B) {
+					runBenchmarkStoreList(ctx, b, store, paginateLimit, maxRevision, c.match, true, nodeNames)
+				})
+			}
+		})
+	})
+	b.Run("Namespace", func(b *testing.B) {
+		for _, c := range cases {
+			b.Run(c.name, func(b *testing.B) {
+				runBenchmarkStoreListNamespace(ctx, b, store, maxRevision, c.match, namespacedNames)
+			})
+		}
+	})
+}
+
+func runBenchmarkStoreListNamespace(ctx context.Context, b *testing.B, store storage.Interface, maxRV int, match metav1.ResourceVersionMatch, namespaceNames []string) {
+	wg := sync.WaitGroup{}
+	objectCount := atomic.Uint64{}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		wg.Add(1)
+		resourceVersion := ""
+		switch match {
+		case metav1.ResourceVersionMatchExact:
+			resourceVersion = fmt.Sprintf("%d", maxRV-99+i%100)
+		case metav1.ResourceVersionMatchNotOlderThan:
+			resourceVersion = "0"
```

Comment:
This may potentially lead to different results - in pathological case we may return from empty watchcache.
I actually suggest using the exact same RV as above.

Reply:
Done